### PR TITLE
Refactor PartialState and AcceleratorState

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR does a *huge* refactor of the `PartialState` device setup to make it easier for new backends to get added, ease the maintenance burden, and centralizes the code based on topics

* When adding a new device, `set_device()` should be updated to include the new backend to be used and calling it
* `_prepare_backend` will handle setting `self.backend` and bringing in any imports that are needed for injecting/monkeypatching (such as `torch_ccl`)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@BenjaminBossan @pacman100 

cc @faaany @huismiling if I could get you to verify on your respective devices this doesn't break anything, it'd be appreciated 🙏 

(As I can only test on CPU/multi-GPU)

### CODE REVIEW

File: src/accelerate/state.py [hunk 1]
```diff
@@ -143,231 +144,118 @@ def __init__(self, cpu: bool = False, **kwargs):
...
-                            torch.xpu.set_device(self.device)
-                    elif is_mlu_available():
-                        self.device = torch.device("mlu", self.local_process_index)
-                        if self.device is not None:
-                            torch.mlu.set_device(self.device)
-                    elif is_npu_available():
-                        self.device = torch.device("npu", self.local_process_index)
-                        if self.device is not None:
-                            torch.npu.set_device(self.device)
+            # Setups up self.backend + imports
```
Comment (BenjaminBossan) [inline src/accelerate/state.py:154]: Please rephrase this comment :D

File: src/accelerate/state.py [hunk 2]
```diff
@@ -143,231 +144,118 @@ def __init__(self, cpu: bool = False, **kwargs):
...
+            self._prepare_backend(cpu, use_sagemaker_dp, kwargs.pop("backend", None))
```
Comment (BenjaminBossan) [inline src/accelerate/state.py:155]: I like the idea of moving this to a sub-method with a limited concern. A suggestion: How about that method returns the `backend` and `distributed_type` instead of setting it on `self`? The code here would change to:

```python
backend, distributed_type = self._prepare_backend(cpu, use_sagemaker_dp, kwargs.pop("backend", None))
self.backend = backend
self.distributed_type = distributed_type
```

That would have the advantage that if I want to understand the code here, I don't need to check `self._prepare_backend` if I'm not interested in these two attributes, as it's immediately obvious that this method is only concerned with determining those.

File: src/accelerate/state.py [hunk 3]
```diff
@@ -143,231 +144,118 @@ def __init__(self, cpu: bool = False, **kwargs):
...
-                    1,
-                )
-                self.local_process_index = local_rank
-                os.environ["RANK"] = str(rank)
-                os.environ["WORLD_SIZE"] = str(size)
-                os.environ["LOCAL_RANK"] = str(local_rank)
-                os.environ["LOCAL_WORLD_SIZE"] = str(local_size)
-
-                if backend == "ccl" and self.distributed_type == DistributedType.MULTI_XPU:
+                    if int(os.environ.get("LOCAL_RANK", -1)) != -1:
```
Comment (BenjaminBossan) [inline src/accelerate/state.py:169]: Let's add a comment here similar to the one above `# Deal with XLA` to delineate the blocks.

File: src/accelerate/state.py [hunk 4]
```diff
@@ -797,6 +685,74 @@ def default_device(self) -> torch.device:
...
+
+    def set_device(self):
+        """
+        Sets the device in `self.device` to the current distributed environment.
+        """
+        if self.device is None:
+            if self.num_processes == 1:
+                self.device = torch.device("cpu") if self._cpu else self.default_device
+            else:
+                device = str(self.distributed_type).split(".")[-1].replace("MULTI_", "").lower()
```
Comment (BenjaminBossan) [inline src/accelerate/state.py:738]: Just a personal preference, feel free to ignore: I prefer early returns:

```python
        if self.device is not None:
            return

        if self.num_processes == 1:
            self.device = torch.device("cpu") if self._cpu else self.default_device
            return

        device = str(self.distributed_type).split(".")[-1].replace("MULTI_", "").lower()
        ...
```

File: src/accelerate/utils/environment.py [hunk 5]
```diff
@@ -176,3 +176,22 @@ def check_fp8_capability():
     """
     cuda_device_capacity = torch.cuda.get_device_capability()
     return cuda_device_capacity >= (8, 9)
+
+
+def get_cpu_distributed_information():
+    """
+    Attempts to return various information about the environment in relation to CPU distributed training
```
Comment (BenjaminBossan) [inline src/accelerate/utils/environment.py:183]: Why "Attempts"? This made me think that this function may raise an error if it can't find the info.

File: src/accelerate/utils/environment.py [hunk 6]
```diff
@@ -176,3 +176,22 @@ def check_fp8_capability():
...
+        ["WORLD_SIZE", "PMI_SIZE", "OMPI_COMM_WORLD_SIZE", "MV2_COMM_WORLD_SIZE"], 1
+    )
+    information["local_rank"] = get_int_from_env(
+        ["LOCAL_RANK", "MPI_LOCALRANKID", "OMPI_COMM_WORLD_LOCAL_RANK", "MV2_COMM_WORLD_LOCAL_RANK"], 0
+    )
+    information["local_world_size"] = get_int_from_env(
+        ["LOCAL_WORLD_SIZE", "MPI_LOCALNRANKS", "OMPI_COMM_WORLD_LOCAL_SIZE", "MV2_COMM_WORLD_LOCAL_SIZE"],
+        1,
+    )
+    return information
```
Comment (BenjaminBossan) [inline src/accelerate/utils/environment.py:197]: Maybe worthwhile to return a `dataclass` instead of a dict so that the caller knows better what it can expect?

File: src/accelerate/state.py [hunk 7]
```diff
@@ -143,231 +144,118 @@ def __init__(self, cpu: bool = False, **kwargs):
...
+                                    "DeepSpeed is not available => install it using `pip3 install deepspeed` or build it from source"
+                                )
+                            from deepspeed import comm as dist
+
+                            if is_xpu_available() and is_ccl_available():
+                                os.environ["CCL_PROCESS_LAUNCHER"] = "none"
+                                os.environ["CCL_LOCAL_SIZE"] = os.environ.get("LOCAL_WORLD_SIZE", "1")
+                                os.environ["CCL_LOCAL_RANK"] = os.environ.get("LOCAL_RANK", "0")
+                            dist.init_distributed(dist_backend=self.backend, auto_mpi_discovery=False, **kwargs)
+                            self.set_device()
```
Comment (BenjaminBossan) [inline src/accelerate/state.py:182]: Just want to ensure that this `set_device()` call is correct here, since it is called again further down below. If it's indeed needed here, maybe add a comment to explain?

Comment (muellerzr) [reply src/accelerate/state.py:182]: It is, we need it to happen before the distributed type is set to deepspeed (and if it's called again after the first time, nothing happens)

File: src/accelerate/state.py (hunk 3 above)
Comment (muellerzr) [reply src/accelerate/state.py:169]: Done

File: src/accelerate/state.py (hunk 2 above)
Comment (muellerzr) [reply src/accelerate/state.py:155]: Makes sense to me, good suggestion.

File: src/accelerate/state.py [hunk 8]
```diff
@@ -797,6 +691,79 @@ def default_device(self) -> torch.device:
...
+                else:
+                    import torch_ccl  # noqa: F401
+
+                backend = "ccl"
+            elif torch.distributed.is_mpi_available():
+                backend = "mpi"
+            else:
+                backend = "gloo"
+        else:
+            distributed_type = DistributedType.NO
```
Comment (huismiling) [inline src/accelerate/state.py:737]: This makes MLU devices `distributed_type = DistributedType.NO`.
distributed_type may be initialized at function head.

**Review by huismiling — COMMENTED** [review]
With distributed_type initialized at _prepare_backend function head, nlp_example.py with MLUs is passed !

File: src/accelerate/state.py (hunk 8 above)
Comment (muellerzr) [reply src/accelerate/state.py:737]: Thanks, I'll look into seeing what I did wrong here. Thought I maintained the behavior but it seems like I did not

Comment (muellerzr) [reply src/accelerate/state.py:738]: Should be fixed now!

File: src/accelerate/utils/environment.py (hunk 6 above)
Comment (muellerzr) [reply src/accelerate/utils/environment.py]: Done, good idea.

I kept it in this class as it's not a useful dataclass outside of here (and not to mention circular imports)

**Comment by faaany** [comment]
Hi @muellerzr , sorry for the late response! I was OOO in the last few days. I tested the patch on XPU for training and it works. Great job! The code looks much cleaner now.

**Comment by muellerzr** [comment]
Final integration tests are being performed here: https://github.com/huggingface/accelerate/actions/runs/8470701075
